plugins/modules/cloud/misc/terraform.py

```diff
@@ -449,12 +480,73 @@ def main():
     if state == 'present' and module.params.get('parallelism') is not None:
         command.append('-parallelism=%d' % module.params.get('parallelism'))
 
+    def format_args(vars):
+        if isinstance(vars, str):
+            return '"{string}"'.format(string=vars)
+        elif isinstance(vars, bool):
+            if vars:
+                return 'true'
+            else:
+                return 'false'
+        elif isinstance(vars, (integer_types, float)):
+            return str(vars)
+        else:
+            return vars
+
+    def process_complex_args(vars):
+        ret_out = []
+        if isinstance(vars, dict):
+            for k, v in vars.items():
+                if isinstance(v, dict):
+                    ret_out.append('{0}={{{1}}}'.format(k, process_complex_args(v)))
+                elif isinstance(v, list):
+                    ret_out.append("{0}={1}".format(k, process_complex_args(v)))
+                elif isinstance(v, (integer_types, float, str, bool)):
+                    ret_out.append('{0}={1}'.format(k, format_args(v)))
+                else:
+                    # only to handle anything unforseen
+                    module.fail_json(msg="Supported types are, dictionaries, lists, strings, integer_types, boolean and float.")
+        if isinstance(vars, list):
+            l_out = []
+            for item in vars:
+                if isinstance(item, dict):
+                    l_out.append("{{{0}}}".format(process_complex_args(item)))
+                elif isinstance(item, list):
+                    l_out.append("{0}".format(process_complex_args(item)))
+                elif isinstance(item, (str, integer_types, float, bool)):
+                    l_out.append(format_args(item))
+                else:
+                    # only to handle anything unforseen
+                    module.fail_json(msg="Supported types are, dictionaries, lists, strings, integer_types, boolean and float.")
+
+            ret_out.append("[{0}]".format(",".join(l_out)))
+        return ",".join(ret_out)
+
     variables_args = []
     for k, v in variables.items():
-        variables_args.extend([
-            '-var',
-            '{0}={1}'.format(k, v)
-        ])
+        if isinstance(v, dict):
+            variables_args.extend([
+                '-var',
+                '{0}={{{1}}}'.format(k, process_complex_args(v))
+            ])
+        elif isinstance(v, list):
+            variables_args.extend([
+                '-var',
+                '{0}={1}'.format(k, process_complex_args(v))
+            ])
+        # terraform does not like shlex_quote fixing double-quotes on the top-level
+        # just passing the plain string and shlex_quote will take care of the rest
```

Comment:
There is no `shlex_quote` involved here, or in the code which uses the result of this evaluation.

Reply:
Sorry if I confused you with the `shlex_quote()`. You're correct we're not using `shlex_quote()` within this module :). When I was digging through why `json_dumps()` didn't work, I stumbled upon the following code in `run_command()`:

```
            # stringify args for unsafe/direct shell usage
            if isinstance(args, list):
                args = b" ".join([to_bytes(shlex_quote(x), errors='surrogate_or_strict') for x in args])
            else:
                args = to_bytes(args, errors='surrogate_or_strict')
```

We're using the List form of args, that's why everything gets wrapped around `shlex_quote()` by default, hence we need not handle anything with regards to strings other than python strings.

This is why `json.dumps()` didn't work in a previous commit.

```
Python 3.9.7 (default, Sep  3 2021, 12:37:55)
[Clang 12.0.5 (clang-1205.0.22.9)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from ansible.module_utils.six.moves import shlex_quote
>>> import json
>>>
>>> rand_str = 'a"b'
>>> shlex_quote(json.dumps(rand_str))
'\'"a\\"b"\''
>>> shlex_quote(rand_str)
'\'a"b\''
```

In summary, we don't need to worry about special charachters in strings, `run_commands()` takes care of it :). I hope this clears the

Reply:
I'm always convinced I'm right all the time, even when I'm wrong :), So just to just to reaffirm myself :). Playbook:

```
---

- hosts: localhost
  gather_facts: false
  tasks:
    - community.general.terraform:
        project_path: "./"
        state: present
        variables:
          dictionaries:
            name: kosala
            age: 44
          list_of_strings:
            - kosala
            - neloufer
            - punu
            - ari
          list_of_objects:
            - name: kosala
              age: 44
            - name: neloufer
              age: 44
            - name: punu
              age: 9
            - name: ari
              age: 1
          boolean: true
          string_type: 'randomstring2"&$%@'
          list_of_lists:
            - [ 1 ]
            - [ 11, 12, 13 ]
            - [ 2 ]
            - [ 3 ]
```

And the result.

```
ok: [localhost] => {
    "changed": false,
    "command": "/usr/local/bin/terraform apply -no-color -input=false -auto-approve -lock=true /var/folders/4g/f8wq3y9s5w9cy92zs0_cs2_h0000gn/T/tmp2v15yiwp.tfplan",
    "invocation": {
        "module_args": {
            "backend_config": null,
            "backend_config_files": null,
            "binary_path": null,
            "check_destroy": false,
            "force_init": false,
            "init_reconfigure": false,
            "lock": true,
            "lock_timeout": null,
            "overwrite_init": true,
            "parallelism": null,
            "plan_file": null,
            "plugin_paths": null,
            "project_path": "./",
            "provider_upgrade": false,
            "purge_workspace": false,
            "state": "present",
            "state_file": null,
            "targets": [],
            "variables": {
                "boolean": true,
                "dictionaries": {
                    "age": 44,
                    "name": "kosala"
                },
                "list_of_lists": [
                    [
                        1
                    ],
                    [
                        11,
                        12,
                        13
                    ],
                    [
                        2
                    ],
                    [
                        3
                    ]
                ],
                "list_of_objects": [
                    {
                        "age": 44,
                        "name": "kosala"
                    },
                    {
                        "age": 44,
                        "name": "neloufer"
                    },
                    {
                        "age": 9,
                        "name": "punu"
                    },
                    {
                        "age": 1,
                        "name": "ari"
                    }
                ],
                "list_of_strings": [
                    "kosala",
                    "neloufer",
                    "punu",
                    "ari"
                ],
                "string_type": "randomstring2\"&$%@"
            },
            "variables_files": null,
            "workspace": "default"
        }
    },
    "outputs": {},
    "state": "present",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "null_resource.mynullresource: Refreshing state... [id=6663805947396045541]\n\nNo changes. Your infrastructure matches the configuration.\n\nTerraform has compared your real infrastructure against your configuration\nand found no differences, so no changes are needed.\n",
    "stdout_lines": [
        "null_resource.mynullresource: Refreshing state... [id=6663805947396045541]",
        "",
        "No changes. Your infrastructure matches the configuration.",
        "",
        "Terraform has compared your real infrastructure against your configuration",
        "and found no differences, so no changes are needed."
    ],
    "workspace": "default"
}
```

Terraform codes https://github.com/kosalaat/terrafrom-tests.git

@felixfontein, is it a practice for us to include functional testing? I couldn't spot any in the existing test cases. TBH, I was not looking hard enough.

Reply:
@kosalaat it is mandatory for new modules, for old ones we usually give it more latitude.

If you have tests ready, by all means please include them - I reckon your functional tests would fall into our "integration" tests mechanism.

Reply:
Please remove / adjust the shlex comment.